File: src/providers/openai-realtime/rules/migrate-beta-to-ga.ts
import { collectOpenAIRealtimeUrlVarNames, findProperty, isOpenAIRealtimeNewWebSocket } from '../utils.js';

const rule = {
  meta: {
    type: 'problem',
    docs: {
      description: 'OpenAI Realtime connections must not pin to the deprecated beta interface',
      category: 'correctness',
      docsUrl: 'https://developers.openai.com/api/docs/guides/realtime',
      rationale:
        "OpenAI's current Realtime guide states that beta integrations must migrate to the GA interface before new work proceeds, and explicitly calls out removing the OpenAI-Beta: realtime=v1 header as a required step. Staying on the beta header keeps a connection locked to the legacy session shape and event names with no confirmed sunset date — a ticking liability rather than an active failure.",
      recommended: true,
    },
    messages: {
      betaHeaderPresent:
        "This OpenAI Realtime connection sends the deprecated 'OpenAI-Beta: realtime=v1' header instead of using the GA interface.",
    },
    schema: [],
  },
  create(context: any) {
    let urlVarNames = new Set<string>();

    return {
      Program(node: any) {
        urlVarNames = collectOpenAIRealtimeUrlVarNames(node);
      },

      NewExpression(node: any) {
        if (!isOpenAIRealtimeNewWebSocket(node, urlVarNames)) return;

        const optionsArg = node.arguments?.[1];
        if (optionsArg?.type !== 'ObjectExpression') return;

        const headersProp = findProperty(optionsArg, 'headers');
        if (headersProp?.value?.type !== 'ObjectExpression') return;

        const betaProp = findProperty(headersProp.value, 'OpenAI-Beta');
        if (!betaProp) return;

        const value = betaProp.value;
        const isRealtimeV1 =
          value?.type === 'Literal' &&
          typeof value.value === 'string' &&
          value.value.includes('realtime=v1');
        if (!isRealtimeV1) return;

        context.report({ node: betaProp, messageId: 'betaHeaderPresent' });
      },
    };
  },
};

export const openaiRealtimeMigrateBetaToGaRule = rule;

File: src/providers/openai-realtime/rules/python/migrate_beta_to_ga.py
"""openai-realtime-migrate-beta-to-ga

Parity with the JS rule: a Realtime connection that sends the deprecated
`OpenAI-Beta: realtime=v1` header must migrate the session/event shapes to
the GA interface instead.
"""
from __future__ import annotations

import ast
import sys
from pathlib import Path

from ast_utils import loc

_PROVIDER = Path(__file__).resolve().parents[2]
if str(_PROVIDER) not in sys.path:
    sys.path.insert(0, str(_PROVIDER))
from utils import (  # noqa: E402
    collect_dict_var_values,
    collect_realtime_url_var_names,
    find_header_value,
    get_headers_arg,
    is_realtime_connect_call,
    literal_string,
    resolve_headers_node,
)

RULE_KEY = "openai-realtime-migrate-beta-to-ga"


def check(tree: ast.AST, path: str, source: str) -> list[dict]:
    url_var_names = collect_realtime_url_var_names(tree)
    dict_var_values = collect_dict_var_values(tree)
    out: list[dict] = []

    for node in ast.walk(tree):
        if not isinstance(node, ast.Call) or not is_realtime_connect_call(node, url_var_names):
            continue
        headers = resolve_headers_node(get_headers_arg(node), dict_var_values)
        pair = find_header_value(headers, "OpenAI-Beta")
        if pair is None:
            continue
        key_node, value_node = pair
        value = literal_string(value_node)
        if value and "realtime=v1" in value:
            line, col, _, _ = loc(key_node)
            out.append({"file": path, "line": line, "column": col, "ruleKey": RULE_KEY})

    return out
